### PR TITLE
Add more Marshal tests

### PR DESCRIPTION
**Pull request by hughbe**

### CODE REVIEW

**Comment by hughbe** [comment]
The test failures are expected:

``` System.Runtime.InteropServices.Tests.ComEventsHelperTests.Remove_NonNullRcw_ThrowsPlatformNotSupportedException
System.Runtime.InteropServices.Tests.ComEventsHelperTests.Combine_NonNullRcw_ThrowsPlatformNotSupportedException
System.Runtime.InteropServices.Tests.ComAwareEventInfoTests.AddEventHandler_DispIdAttribute_ThrowsPlatformNotSupportedException
System.Runtime.InteropServices.MarshalTests.SetComObjectData
System.Runtime.InteropServices.MarshalTests.GetComObjectData
```
These used to throw PNSE but now do not. How do I go about fixing this? I'm slightly confused as the tests will fail in corefx until coreclr is updated!

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/Common/CommonTypes.Windows.cs [hunk 1]
```diff
@@ -2,12 +2,34 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#pragma warning disable CS0618 // Type or member is obsolete
+
 namespace System.Runtime.InteropServices.Tests.Common
 {
     [ComImport]
     [Guid("927971f5-0939-11d1-8be1-00c04fd8d503")]
```
Comment (AaronRobinsonMSFT) [inline src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/Common/CommonTypes.Windows.cs]: I missed this when it was checked in, but with this change it is much more obvious. The `Guid` values should be unquie for all interfaces.

@luqunl what is the behavior if multiple interfaces have the same guid? Is this basically ignored by the type system unless a QI is done?

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetComObjectDataTests.cs [hunk 2]
```diff
@@ -6,25 +6,18 @@
 
 namespace System.Runtime.InteropServices.Tests
 {
-    public class GetComObjectDataTests
+    public partial class GetComObjectDataTests
     {
         [Fact]
-        [SkipOnTargetFramework(~TargetFrameworkMonikers.NetFramework, "Marshal.GetComObjectData is not implemented in .NET Core.")]
-        public void GetComObjectData_NetFramework_ReturnsExpected()
```
Comment (luqunl) [inline src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetComObjectDataTests.cs]: why do you remove this scenario?

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/Common/CommonTypes.Windows.cs (hunk 1 above)
Comment (hughbe) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/Common/CommonTypes.Windows.cs]: Do you suggest I change the GUID? I actually don't really know a very good list of GUIDs that re valid... any ideas?

Comment (AaronRobinsonMSFT) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/Common/CommonTypes.Windows.cs]: If you are on Windows, you can generate them quickly from the Visual Studio Developer command prompt. The following command will generate 10 new GUIDs: `uuidgen /c /n10`

Comment (luqunl) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/Common/CommonTypes.Windows.cs]: I would prefer different guid for different types unless there is a special scenario which require to use same guid.

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/ZeroFreeGlobalAllocUnicodeTests.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace System.Runtime.InteropServices.Tests
+{
+    public class ZeroFreeGlobalAllocUnicodeTests
+    {
+        [Fact]
+        public void ZeroFreeGlobalAllocUnicode_ValidPointer_Success()
+        {
+            IntPtr ptr = Marshal.StringToHGlobalUni("hello");
+            Marshal.ZeroFreeGlobalAllocUnicode(ptr);
```
Comment (luqunl) [inline src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/ZeroFreeGlobalAllocUnicodeTests.cs]: this is incorrect. memory allocated by Marshal.StringToHGlobalUni should be released by Marshal.FreeHGlobal

Comment (luqunl) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/ZeroFreeGlobalAllocUnicodeTests.cs]: similar to others. memory allocated by SecureStringToGlobalAllocUnicode should be released by Marshal.ZeroFreeGlobalAllocUnicode

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/ThrowExceptionForHRTests.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        {
+            yield return new object[] { unchecked((int)0x80020006), IntPtr.Zero };
+            yield return new object[] { unchecked((int)0x80020101), IntPtr.Zero };
+            yield return new object[] { unchecked((int)0x80020006), (IntPtr)(-1) };
+            yield return new object[] { unchecked((int)0x80020101), (IntPtr)(-1) };
+        }
+
+        [Theory]
+        [MemberData(nameof(ThrowExceptionForHR_ErrorInfo_TestData))]
+        [ActiveIssue(30866)]
```
Comment (luqunl) [inline src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/ThrowExceptionForHRTests.cs]: 30866 is fixed(or wokaround). You need to call another Marshal.GetExceptionForHR to clear IErrorInfo in that thread. see https://github.com/dotnet/corefx/pull/31136

**Comment by hughbe** [comment]
@luquni sorry to be a pain but could you actually review the other open pRs first as they contain things split off from this PR?

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/ZeroFreeGlobalAllocUnicodeTests.cs (hunk 3 above)
Comment (hughbe) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/ZeroFreeGlobalAllocUnicodeTests.cs]: See https://github.com/dotnet/corefx/pull/31151/files

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/ThrowExceptionForHRTests.cs (hunk 4 above)
Comment (AaronRobinsonMSFT) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/ThrowExceptionForHRTests.cs]: @luqunl I think this should be addressed when @hughbe merges - seea754a612eed6d16b52d7f75c602f71eb367760ae.

Comment (hughbe) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/ThrowExceptionForHRTests.cs]: Yes. I'm going to wait on solving merge conflicts in this PR till my open `Marshal` PRs are merged (https://github.com/dotnet/corefx/pulls/hughbe)

**Comment by karelz** [comment]
@luqunl @hughbe what is the plan for this PR?

**Comment by AaronRobinsonMSFT** [comment]
My understanding was that @hughbe was breaking up this PR into smaller ones to help make reviewing easier. @hughbe is that still the plan? Can we close this PR?

**Comment by hughbe** [comment]
Ummm this shouldn’t be closed but needs to be rebased! I’m not with a laptop for the next 3 weeks so will do when I’m back. Is that OK?

**Comment by karelz** [comment]
Let's close it for now to avoid noise in opened PRs.
You can reopen it (and rebase) in 3 weeks when you have time to finish it. Will that work for you @hughbe?

**Comment by hughbe** [comment]
Yeah go on

**Pull request closed without merging**